### PR TITLE
Alloc. awareness should ignore nodes without attr

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we count `null` (i.e. missing) as a valid attribute value in
allocation awareness, even though allocation awareness forbids the
allocation of shards to such a node. Prior to #69334 this didn't matter,
a data node without allocation attributes was pointless.

However, #69334 means we now can allocate shards to such a node: for
instance, there is no need for nodes holding only enrich indices to have
allocation attributes. Therefore we should stop counting `null` as one
of the attribute values.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java [hunk 1]
```diff
@@ -904,4 +905,35 @@ public void testDisabledByAutoExpandReplicas() {
...
+                                .build())
+                        .nodes(DiscoveryNodes.builder()
+                                .add(newNode("A-0", singletonMap("zone", "a")))
+                                .add(newNode("A-1", singletonMap("zone", "a")))
+                                .add(newNode("B-0", singletonMap("zone", "b")))
+                                .add(newNode("B-1", singletonMap("zone", "b")))
+                                .add(newNode("X-0", emptyMap()))
+                        ).build(), strategy);
+
+        assertThat(clusterState.getRoutingNodes().shardsWithState(UNASSIGNED), empty());
```
Comment (ywelsch) [inline server/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java:936]: where do we assert that `X-0` is ignored, i.e. does not hold a shard copy?

**Review by ywelsch — COMMENTED** [review]
I find it a bit difficult to understand the implications of this PR. will it break clusters that have `cluster.routing.allocation.awareness.attributes` defined but where the nodes are missing the attributes?

**Comment by DaveCTurner** [comment]
Today if you have a cluster containing a data node without all the awareness attributes defined then those nodes get no shards at all.

https://github.com/elastic/elasticsearch/blob/bb3ea99850e16a06d77649712595bcf45964250c/server/src/main/java/org/elasticsearch/cluster/routing/allocation/decider/AwarenessAllocationDecider.java#L142-L145

It's hard to imagine that there are any clusters in this state today, they would have these empty data nodes and would almost always have unassigned shards too.

The effect of this change is to reduce the effective number of attribute values in such a cluster, which relaxes the upper bound on how many shards can be allocated to each value, which shouldn't break anything.

File: server/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/test/java/org/elasticsearch/cluster/routing/allocation/AwarenessAllocationTests.java:936]: Added in d7216bb7f0223d1cc92ee105f5e6137520c8fd3c

**Review by ywelsch — APPROVED** [review]
ok, I see, thank you for the explanation. LGTM

**Pull request merged**